### PR TITLE
CDAP-12500 correct run records in starting state

### PR DESCRIPTION
**Pull request by albertshau**
Now that there is a starting state, the run record corrector
needs to correct run records that are either in the running
state or in the starting state.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java [hunk 1]
```diff
@@ -83,17 +85,21 @@ private void validateAndCorrectRunningRunRecords(final ProgramType programType,
...
-                      @Override
-                      public boolean apply(RunRecordMeta input) {
-                        String runId = input.getPid();
-                        // Check if it is not actually running.
-                        return !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
-                      }
-                    }).values();
+    List<RunRecordMeta> notActuallyRunning = new ArrayList<>();
+
+    notActuallyRunning.addAll(store.getRuns(ProgramRunStatus.RUNNING, filter).values());
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java:99]: should we also check SUSPENDED?

**Comment by sreevatsanraman** [comment]
LGTM

**Comment by sreevatsanraman** [comment]
Let's merge if the checkstyle passes and tested on a cluster.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java (hunk 1 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java:99]: +1 It is a good idea to check all non-final states

Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java:99]: scenario to reproduce would be similar: do a `yarn application -kill <app-id>` on a suspended workflow.

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java:99]: Related to that, it seems that the transition table does not allow going from RESUMING to FAILED. https://github.com/caskdata/cdap/blob/release/4.3/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java#L107

Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java:99]: I'm going to open a separate jira for suspended correction to limit the changes this close to the release. Suspended is an existing bug and I would imagine its far less common. I didn't even know you could suspend a workflow until now.

**Comment by albertshau** [comment]
checkstyle passed, verified that it corrected the run record on a cluster.

**Pull request merged**